src/axom/core/execution/internal/cuda_exec.hpp

```diff
@@ -0,0 +1,140 @@
+// Copyright (c) 2017-2019, Lawrence Livermore National Security, LLC and
+// other Axom Project Developers. See the top-level COPYRIGHT file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+
+#ifndef AXOM_CUDA_EXEC_HPP_
+#define AXOM_CUDA_EXEC_HPP_
+
+#include "axom/config.hpp"
+#include "axom/core/memory_management.hpp"
+
+#include "RAJA/RAJA.hpp"
+#include "umpire/Umpire.hpp"
+
+#ifndef RAJA_ENABLE_CUDA
+#error *** CUDA_EXEC requires a CUDA enabled RAJA ***
+#endif
+
+#if !defined(UMPIRE_ENABLE_CUDA) && !defined(UMPIRE_ENABLE_UM)
+#error *** CUDA_EXEC requires a CUDA enabled UMPIRE with UM support ***
+#endif
+
+namespace axom
+{
+
+enum ExecutionMode
+{
+  SYNCHRONOUS,
+  ASYNC
+};
+
+/*!
+ * \brief Indicates parallel execution on the GPU with CUDA.
+ *
+ * \tparam BLOCK_SIZE the number of CUDA threads in a block.
+ * \tparam ExecutionMode indicates synchronous or asynchronous execution.
+ */
+template < int BLOCK_SIZE, ExecutionMode EXEC_MODE=SYNCHRONOUS >
+struct CUDA_EXEC{ };
+
+/*!
+ * \brief execution_space traits specialization for CUDA_EXEC.
+ *
+ * \tparam BLOCK_SIZE the number of CUDA threads to launch
+ *
+ */
+template < int BLOCK_SIZE >
+struct execution_space< CUDA_EXEC< BLOCK_SIZE, SYNCHRONOUS > >
+{
+  using loop_policy   = RAJA::cuda_exec< BLOCK_SIZE >;
+
+  /* *INDENT-OFF* */
+  using loop2d_policy =
```

Comment:
Are these policy definitions too constraining, potentially? They might be suboptimal depending on the kernel.

Reply:
The loop 2d/3d policies are intended for generic i-j-k loops. Is there something better and generic that we should do in this case instead?

These type of policies should not be used for things like, loop over the DOFs of each element and process each DOF independently.  Indeed, that would be suboptimal in this case. My expectation is that for specific use cases, different policies will be defined at the application level. 

That said, exposing these loop policies here, may be misleading application developers to use them regardless.

Perhaps, we shouldn't define these policies here at all. We could leave it to the application, perhaps, by inheriting and extending the `execution_space` traits class here. Thoughts?

Reply:
That last part seems better, I think.

Reply:
Sounds good -- I'll remove the predefined kernel policies from here.